File: Landmark-Kazemi/c++/kazemi-demo.cpp
/*----------------------------------------------
 * Usage:
 * kazemi-demo <face_cascade_model> <kazemi_model> <testing_images>
 *
 * Example:
 * kazemi-demo models/face_cascade.xml models/model.dat test/images_test.txt
 *
* Notes:
 * the user should provides the list of testing images in a text file
 * example of contents for images_test.txt:
 * ../trainset/image_0001.png
 * ../trainset/image_0002.png
 *--------------------------------------------------*/

#include <stdio.h>
#include <fstream>
#include <sstream>
#include "opencv2/face.hpp"
#include "opencv2/highgui.hpp"
#include "opencv2/imgcodecs.hpp"
#include "opencv2/videoio.hpp"
#include "opencv2/objdetect.hpp"
#include "opencv2/imgproc.hpp"

#include <iostream>
#include <string>
#include <vector>
#include <ctime>

using namespace std;
using namespace cv;
using namespace cv::face;

bool myDetector(InputArray image, OutputArray faces, CascadeClassifier *face_cascade)
{
    Mat gray;
    if (image.channels() > 1)
        cvtColor(image, gray, COLOR_BGR2GRAY);
    else
        gray = image.getMat().clone();

    equalizeHist(gray, gray);

    vector<Rect> faces_;
    face_cascade->detectMultiScale(gray, faces_, 1.4, 2, CASCADE_SCALE_IMAGE, Size(30, 30));
    Mat(faces_).copyTo(faces);
    return true;
}

bool parseArguments(int argc, char** argv,
    String & cascade,
    String & kazemi_model,
    String & test_images
){
    const String keys =
        "{ @f face-cascade    |      | (required) path to the cascade model file for the face detector }"
        "{ @m model-path      |      | (required) path to the trained Kazemi model}"
        "{ @t test-images     |      | (required) path of a text file contains the list of paths to the test images}"
        "{ help h usage ?     |      |  kazemi-webcam-demo -face-cascade  -model -test-images\n"
             " example: kazemi-webcam-demo models/face_cascade.xml  models/face_landmark_model.dat test/images_test.txt }"
    ;
    CommandLineParser parser(argc, argv, keys);
    parser.about("hello");

    if (parser.has("help")){
        parser.printMessage();
        return false;
    }

    cascade = String(parser.get<string>("face-cascade"));
    kazemi_model = String(parser.get<string>("model-path"));
    test_images = String(parser.get<string>("test-images"));

    if(cascade.empty() || kazemi_model.empty()){
        cerr << "one or more required arguments are not found" << '\n';
        cout<<"face-cascade : "<<cascade.c_str()<<endl;
        cout<<"kazemi-model : "<<kazemi_model.c_str()<<endl;
        cout<<"test_images : "<<test_images.c_str()<<endl;
        parser.printMessage();
        return false;
    }
    return true;
}


int main(int argc, char** argv )
{
    String cascade_path, kazemi_model, test_images_path;
    if(!parseArguments(argc, argv, cascade_path, kazemi_model, test_images_path))
       return -1;

    /* Create the Facemark Instance */
    CascadeClassifier face_cascade;
    face_cascade.load(cascade_path);
    FacemarkKazemi::Params params;
    Ptr<FacemarkKazemi> facemark = FacemarkKazemi::create(params);
    facemark->setFaceDetector((FN_FaceDetector)myDetector, &face_cascade);
    facemark->loadModel(kazemi_model);

    /* Loading the testing dataset */
    String testFiles, testPts;
    testFiles = test_images_path;
    testPts = test_images_path; //unused
    vector<String> images;
    vector<String> facePoints;
    loadDatasetList(testFiles, testPts, images, facePoints);

    vector<Rect> faces;
    for(int i=0; i<(int)images.size(); i++){
        vector<vector<Point2f> > landmarks;
        Mat img = imread(images[i]);
        facemark->getFaces(img, faces);
        
        if(faces.size() > 0){
            double newtime = (double)getTickCount();
            facemark->fit(img, faces, landmarks);
            double fittime = ((getTickCount() - newtime)/getTickFrequency());
            for(unsigned j=0;j<faces.size();j++){
                drawFacemarks(img, landmarks[j], Scalar(0,0,255));
                rectangle(img, faces[j], Scalar(255,0,255));
            }
            printf("Time taken - %f ms\n",fittime*1000);
            imshow("Landmark Image", img);;
            char k = waitKey(0);
            if (k == 27)
              break;
        }else{
            cout<<"Face not found in the Image"<<endl;
        }
    }
}

File: Landmark-Kazemi/python/kazemi-demo.py
import os
import time
import cv2
import argparse

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Demonstration of kazemi facial landmark algorithm.')
    parser.add_argument('--face_cascade', type=str, help="Path to the cascade model file for the face detector",
                        default=os.path.join(os.path.dirname(os.path.realpath(__file__)),'models','lbpcascade_frontalface_improved.xml'))
    parser.add_argument('--kazemi_model', type=str, help="Path to the kazemi trained model file",
                        default=os.path.join(os.path.dirname(os.path.realpath(__file__)),'models','face_landmark_model.dat'))
    parser.add_argument('--test_images', type=str, help="Path of the file containing the test images", default=os.path.join(os.path.dirname(os.path.realpath(__file__)),'test','images_test.txt'))
    parser.add_argument('--verbose', action='store_true')
    args = parser.parse_args()

    face_cascade = cv2.CascadeClassifier(args.face_cascade)
    facemark = cv2.face.createFacemarkKazemi()
    facemark.loadModel(args.kazemi_model)

    start = time.time()
    status, images_test, _ = cv2.face.loadDatasetList(args.test_images, args.test_images)
    assert(status == True)
    if args.verbose:
        print("Loading the dataset took {} seconds".format(time.time()-start))

    for image in images_test:
        frame = cv2.imread(image)
        if frame is not None :
            gray_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
            faces = face_cascade.detectMultiScale(gray_frame, 1.3, 5)
            if len(faces) > 0:
                start = time.time()
                status, landmarks = facemark.fit(frame, faces)
                if args.verbose:
                    print("status : ", status)
                    print("landmarks : ", landmarks)
                    print("time taken to fit : {0:.2f} ms".format(1000*(time.time() - start)))
                for f in range(len(landmarks)):
                    cv2.face.drawFacemarks(frame, landmarks[f])
            else:
                print("No Face Detected")

            cv2.imshow('Landmark_Window', frame)
            if cv2.waitKey(0) == 27:
              break;
        else:
            break
    cv2.destroyAllWindows()
